libbpf-tools/trace_helpers.c

```diff
@@ -169,6 +175,604 @@ const struct ksym *ksyms__get_symbol(const struct ksyms *ksyms,
+	if (unlink(tmpfile) == -1)
+		fprintf(stderr, "failed to unlink %s: %s\n", tmpfile,
+			strerror(errno));
+	if (write(fd, image, sz) == -1) {
+		fprintf(stderr, "failed to write to vDSO image: %s\n",
+			strerror(errno));
+		close(fd);
+		fd = -1;
+		goto err_out;
+	}
+
+err_out:
+	fclose(f);
+	free(image);
+	return fd;
+}
+
+static int dso__load_sym_table_from_vdso_image(struct dso *dso)
+{
+	int fd = create_tmp_vdso_image(dso);
+
+	if (fd < 0)
+		return -1;
+	return dso__load_sym_table_from_elf(dso, fd);
+}
+
+static int dso__load_sym_table(struct dso *dso)
+{
+	if (dso->type == UNKNOWN)
+		return -1;
+	if (dso->type == PERF_MAP)
+		return dso__load_sym_table_from_perf_map(dso);
+	if (dso->type == EXEC || dso->type == DYN)
+		return dso__load_sym_table_from_elf(dso, 0);
+	if (dso->type == VDSO)
+		return dso__load_sym_table_from_vdso_image(dso);
+	return -1;
+}
+
+static struct sym *dso__find_sym(struct dso *dso, uint64_t offset)
+{
+	unsigned long sym_addr;
+	int start, end, mid;
+
+	if (!dso->syms && dso__load_sym_table(dso))
+		return NULL;
+
+	start = 0;
+	end = dso->syms_sz - 1;
+
+	/* find largest sym_addr <= addr using binary search */
+	while (start < end) {
+		mid = start + (end - start + 1) / 2;
+		sym_addr = dso->syms[mid].start;
+
+		if (sym_addr <= offset)
+			start = mid;
+		else
+			end = mid - 1;
+	}
+
+	if (start == end && dso->syms[start].start <= offset)
+		return &dso->syms[start];
+	return NULL;
+}
+
+struct syms *syms__load(pid_t tgid)
+{
+	char buf[PATH_MAX], perm[5];
+	struct syms *syms;
+	char fname[128];
+
+	struct map map;
+	char *name;
+	FILE *f;
+	int ret;
+
+	snprintf(fname, sizeof(fname), "/proc/%ld/maps", (long)tgid);
+	f = fopen(fname, "r");
+	if (!f)
+		return NULL;
+
+	syms = calloc(1, sizeof(*syms));
+	if (!syms)
+		goto err_out;
+	syms->tgid = tgid;
+
+	while (true) {
+		ret = fscanf(f, "%lx-%lx %4s %lx %lx:%lx %lu%[^\n]",
+			     &map.start_addr, &map.end_addr, perm,
+			     &map.file_off, &map.dev_major,
+			     &map.dev_minor, &map.inode, buf);
+		if (ret == EOF && feof(f))
+			break;
+		if (ret != 8)	/* perf-<PID>.map */
+			goto err_out;
+
+		if (perm[2] != 'x')
+			continue;
+
+		name = buf;
+		while (isspace(*name))
+			name++;
+		if (!is_file_backed(name))
+			continue;
+
+		if (syms__add_dso(syms, &map, name))
+			goto err_out;
+	}
+
+	fclose(f);
+	return syms;
+
+err_out:
+	syms__free(syms);
+	fclose(f);
+	return NULL;
+}
+
+void syms__free(struct syms *syms)
+{
+	int i;
+
+	if (!syms)
+		return;
+
+	for (i = 0; i < syms->dso_sz; i++)
+		dso__free_fields(&syms->dsos[i]);
+	free(syms->dsos);
+	free(syms);
+}
+
+const struct sym *syms__map_addr(const struct syms *syms, unsigned long addr)
+{
+	struct dso *dso;
+	uint64_t offset;
+
+	dso = syms__find_dso(syms, addr, &offset);
+	if (!dso)
+		return NULL;
+	return dso__find_sym(dso, offset);
+}
+
+const struct sym *syms__get_symbol(const struct syms *syms,
+				   const char *name)
+{
+	return NULL;
+}
+
+struct syms_vec {
```

Comment:
I think syms_vec is misleading name. This is a cache of per-file/per-process symbols, right? It doesn't have to be internally implemented as a vector (dynamic array). So maybe syms_cache or something along those lines?

Reply:
Sure, fixed, thanks!